docs: rework lifecycle hooks topic for style and clarity (#36353)
angular/angular: aio/content/guide/lifecycle-hooks.md

> # Managing the component lifecycle

Author: "Hooking into the component lifecycle"

> # Managing the component lifecycle

Reviewer: I like that!

A component instance has a lifecycle that starts when Angular instantiates the component class and renders the component view along with its child views.
The lifecycle continues with change detection.
As Angular checks when data-bound properties change, updating both the view and the instance.
The lifecycle ends when Angular destroys the component instance and removes its rendered template from the DOM.
Directives have a similar lifecycle, as Angular creates, updates, and destroys instances in the course of execution.

Your application can use [lifecycle hook methods](guide/glossary#lifecycle-hook "Definition of lifecycle hook") to tap into key events in the lifecycle of a component or directive in order to initialize new instances, initiate change detection when needed, respond to updates during change detection, and clean up before deletion of instances.

> #### Prerequisites

Reviewer 2: I think Prerequisites should be level 2. I think it makes it more visible to the user--especially when they are skimming the headings.

```suggestion
## Prerequisites
```

    </td>
    <td>

      Initialize the directive or component after Angular first displays the data-bound properties
      and sets the directive or component's input properties.

>       See details in [Initializing on instance creation](#oninit) in this document.

Reviewer: Might want to update the title here too

```suggestion
      See details in [Initializing a component or directive](#oninit) in this document.
```

    </td>
    <td>

      Cleanup just before Angular destroys the directive or component.
      Unsubscribe Observables and detach event handlers to avoid memory leaks.
      See details in [Cleaning up on instance destruction](#ondestroy) in this document.

    </td>

    <td>

>       Called  immediately before Angular destroys the directive or component.

Reviewer 3: It looks like there's an extra space after "called", but it doesn't look like it in the suggestion. **rubs eyes**

Author: There was 👍

    </td>
  </tr>
</table>

{@a the-sample}

### Lifecycle example set

The <live-example></live-example>
demonstrates the use of lifecycle hooks through a series of exercises
presented as components under the control of the root `AppComponent`.
In each case a *parent* component serves as a test rig for
a *child* component that illustrates one or more of the lifecycle hook methods.

> The following table lists the exercises with brief descriptions. The sample code is also used to illustrate specific tasks in the following sections.

Reviewer 3: ```suggestion
The following table lists the exercises with brief descriptions. 
The sample code is also used to illustrate specific tasks in the following sections.
```

(hard return for new sentence)

</div>

The sequence of log messages follows the prescribed hook calling order:
`OnChanges`, `OnInit`, `DoCheck`&nbsp;(3x), `AfterContentInit`, `AfterContentChecked`&nbsp;(3x),
`AfterViewInit`, `AfterViewChecked`&nbsp;(3x), and `OnDestroy`.

<div class="alert is-helpful">

  Notice that the log confirms that input properties (the `name` property in this case) have no assigned values at construction.
  The input properties are available to the `onInit()` method for further initialization.

</div>

Had the user clicked the *Update Hero* button, the log would show another `OnChanges` and two more triplets of
`DoCheck`, `AfterContentChecked` and `AfterViewChecked`.
Notice that these three hooks fire *often*, so it is important to keep their logic as lean as possible.

{@a spy}

### Use directives to watch the DOM

The `Spy` example demonstrates how you can use hook method for directives as well as components.
The `SpyDirective` implements two hooks, `ngOnInit()` and `ngOnDestroy()`, in order to discover when a watched element is in the current view.

This template applies the `SpyDirective` to a `<div>` in the `ngFor` *hero* repeater managed by the parent `SpyComponent`.

> The example does not perform any initialization or clean-up. It just tracks the appearance and disappearance of an element in the view by recording when the directive itself is instantiated and destroyed.

Reviewer 3: Hard return for new sentence, "It just tracks..."

</div>

The log entries appear as the string value of the *power* property changes.
Notice, however, that the `ngOnChanges()` method does not catch changes to `hero.name`.
This is because Angular calls the hook only when the value of the input property changes.
In this case, `hero` is the input property, and the value of the `hero` property is the *reference to the hero object*.
The object reference did not change when the value of its own `name` property changed.

{@a afterview}

### Responding to view changes

> As the Angular traverses the [view hierarchy](guide/glossary#view-hierarchy "Definition of view hierarchy definition") during change detection, it needs to be sure that a change in a child does not attempt to cause a change in its own parent. Such a change would not be rendered properly, because of the [unidirectional data flow rule](guide/glossary#unidirectional-data-flow "Definition").

Reviewer: This isn't a rule (it's just how the code runs).

```suggestion
As Angular traverses the [view hierarchy](guide/glossary#view-hierarchy "Definition of view hierarchy definition") during change detection, it needs to be sure that a change in a child does not attempt to cause a change in its own parent. Such a change would not be rendered properly, because of how [unidirectional data flow](guide/glossary#unidirectional-data-flow "Definition") works.
```

</div>

The log entries appear as the string value of the *power* property changes.
Notice, however, that the `ngOnChanges()` method does not catch changes to `hero.name`.
This is because Angular calls the hook only when the value of the input property changes.
In this case, `hero` is the input property, and the value of the `hero` property is the *reference to the hero object*.

> The object reference did not change when the value of its own `name` property changed.

Author: It is.

</div>

The log entries appear as the string value of the *power* property changes.
Notice, however, that the `ngOnChanges()` method does not catch changes to `hero.name`.
This is because Angular calls the hook only when the value of the input property changes.
In this case, `hero` is the input property, and the value of the `hero` property is the *reference to the hero object*.
The object reference did not change when the value of its own `name` property changed.

{@a afterview}

### Responding to view changes

> As the Angular change detector traverses the [view hierarchy](guide/glossary#view-hierarchy "Definition of view hierarchy definition"), it needs to be sure that a change in a child does not cause a change in its own parent that would lead to an infinite loop. To ensure this, it applies the [unidirectional data flow rule](guide/glossary#unidirectional-data-flow "Definition").

Reviewer: Angular doesn't really have a separate change detector entity as such

```suggestion
As Angular traverses the [view hierarchy](guide/glossary#view-hierarchy "Definition of view hierarchy definition") during change detection, it needs to be sure that a change in a child does not attempt to cause a change in its own parent, which would not be rendered properly due to unidirectional data flow](guide/glossary#unidirectional-data-flow "Definition"). 
```

> This rule forbids updates to a view after it has been composed.

Reviewer: ```suggestion

```

> If you need to make a change that would violate this rule, you must include a delay, so that the change to the parent is not made until the following change-detection cycle.
> The examples illustrate how to make such changes safely.

Reviewer: ```suggestion
If you need to make a change that inverts the expected data flow, it's necessary to trigger a new change detection cycle so the change is rendered.
The examples illustrate how to make such changes safely.
```

Be very careful about how much logic or computation you put into one of these methods.

<div class="lightbox">

  <img src='generated/images/guide/lifecycle-hooks/after-view-anim.gif' alt="AfterView">

</div>

{@a aftercontent}
{@a aftercontent-hooks}
{@a content-projection}

### Responding to projected content changes

*Content projection* is a way to import HTML content from outside the component and insert that content
into the component's template in a designated spot.
You can identify content projection in a template by looking for the following constructs.

  * HTML between component element tags.
  * The presence of `<ng-content>` tags in the component's template.

<div class="alert is-helpful">

  AngularJS developers know this technique as *transclusion*.

</div>

The *AfterContent* sample explores the `AfterContentInit()` and `AfterContentChecked()` hooks that Angular calls *after* Angular projects external content into the component.

Consider this variation on the [previous _AfterView_](#afterview) example.
This time, instead of including the child view within the template, it imports the content from

> the `AfterContentComponent`'s parent. Here's the parent's template:

Reviewer 4: "The following is the parent's template:"

which can only be reached by querying for the child view via the property decorated with
[@ViewChild](api/core/ViewChild).

<code-example path="lifecycle-hooks/src/app/after-view.component.ts" region="hooks" header="AfterViewComponent (class excerpts)"></code-example>

{@a wait-a-tick}

#### Wait before updating the view

In this example, the `doSomething()` method updates the screen when the hero name exceeds 10 characters, but waits a tick before updating `comment`.

<code-example path="lifecycle-hooks/src/app/after-view.component.ts" region="do-something" header="AfterViewComponent (doSomething)"></code-example>

> This is an example of how to abide by Angular's *unidirectional data flow* rule, which forbids updates to the view after it has been composed.

Reviewer: Calling unidirectional data flow a "rule" makes it sound like an arbitrary stylistic thing that we enforce. In reality, it's just the order that components are checked, which is necessary to understand if you want your components to render as expected.

```suggestion
```

Both the `AfterViewInit()` and `AfterViewChecked()` hooks fire after the component's view has been composed.
If you modify the code so that the hook updates the component's data-bound `comment` property immediately, you can see that Angular throws an error.

The `LoggerService.tick_then()` statement postpones the log update

> for one turn of the browser's JavaScript cycle and that's just long enough.

Reviewer: The docs make it sound like it's a matter of waiting long enough, but it's really that causing a browser tick in itself *causes* change detection to run (because of how zone.js works)

```suggestion
for one turn of the browser's JavaScript cycle, which triggers a new change detection cycle to run.
```

Angular calls both *AfterContent* hooks before calling either of the *AfterView* hooks.
Angular completes composition of the projected content *before* finishing the composition of this component's view.
There is a small window between the `AfterContent...` and `AfterView...` hooks that allows you to modify the host view.

</div>

{@a docheck}

## Defining custom change detection

To monitor changes that occur where `ngOnChanges()` won't catch them, you can implement your own change check, as shown in the *DoCheck* example.
This example shows how you can use the `ngDoCheck()` hook to detect and act upon changes that Angular doesn't catch on its own.

The *DoCheck* sample extends the *OnChanges* sample with the following `ngDoCheck()` hook:

<code-example path="lifecycle-hooks/src/app/do-check.component.ts" region="ng-do-check" header="DoCheckComponent (ngDoCheck)"></code-example>

This code inspects certain _values of interest_, capturing and comparing their current state against previous values.
It writes a special message to the log when there are no substantive changes to the `hero` or the `power`

> so you can see how often `DoCheck` is called. The results are illuminating:

Reviewer 3: ```suggestion
so you can see how often `DoCheck()` is called. The results are illuminating:
```

{@a no-unidirectional-flow-worries}

<div class="alert is-helpful>

<header>No need to wait for content updates</header>

This component's `doSomething()` method updates the component's data-bound `comment` property immediately.

> There's no need to delay the update to avoid violating by Angular's [unidirectional data flow rule](#wait-a-tick).

Reviewer: ```suggestion
There's no need to delay the update to [ensure proper rendering](#wait-a-tick).
```
